README.md
# JVM Platform

[![Maven Central](https://img.shields.io/maven-central/v/com.fleshgrinder/jvm-platform)][Maven Central]
[![GitHub CI Workflow](https://img.shields.io/github/workflow/status/Fleshgrinder/jvm-platform/ci)](https://github.com/Fleshgrinder/jvm-platform/actions)
[![Code Coverage](https://img.shields.io/codecov/c/github/Fleshgrinder/jvm-platform)][CodeCov]

**JVM Platform** provides type-safe access to the Java Virtual Machine
(JVM) `os.name` and `os.arch` system properties with tested support for a wide
range of operating systems and architectures. It also provides standardized
platform identifiers that can be trusted and dependent upon, unlike the system
properties that are provided by the JVM. Additional optional support to further
refine a platform or detect features is available (e.g., detect [musl]). The
library is Java 1.8+ compatible, has zero dependencies, does not allocate
anything unless asked, features proper nullability information for perfect IDE
and Kotlin compatibility, and is rigorously tested (2,000+ test cases).

- **Install** from [Maven Central]
- **Usage** is [illustrated here][Usage.java]

## Motivation

There are many libraries out there that provide the ability to retrieve platform
information but all of them come along with a lot of additional baggage that is
not required. Most of them also allocate data globally throughout the entire
lifetime of the program, for no good reason since the information is most often
required only at certain points during execution to make a decision. The
allocation additionally makes little sense because the computation of the
information is extremely lightweight and probably never the bottleneck in any
serious program.

```java
final Platform platform = Platform.current();
System.out.println(platform.os);   // e.g. LINUX
System.out.println(platform.arch); // e.g. X86_64
System.out.println(platform);      // e.g. linux-x86-64
```

A feature that most libraries out there are missing is the ability to determine
the C ABI of the current JVM. [musl] is getting more and more important in a
world with containers and the JVM has support for it since a while, however, it
is not possible to find out if the current JVM was built against musl or glibc,
this complicates things when we have to make decisions on the kind of native
components we can use.

```java
if (Env.current() == Env.MUSL) {
    // should be a musl system
}
```

A feature that is missing from all libraries we found is the ability to parse
any kind of string to determine the platform. Something that is important when
dealing with native software where we might want to parse a filename or path and
get the corresponding information. A good use case here is for instance when we
want to redistribute some native executable that is published somewhere to a
Maven repository with Gradle (see [gradle-exe-plugin]).

```java
final Platform platform = Platform.parse("OpenJDK11U-jdk_x64_linux_hotspot_11.0.11_9.tar.gz");
assert platform.os == LINUX;
assert platform.arch == X86_64;
```

## Project Info

- Licensed under the [Unlicense], see [UNLICENSE.md] for details.
- Contributions are highly appreciated, see [CONTRIBUTING.md] for details.
- We use [semantic versioning] and [keep a changelog], available versions and
  changes are listed on our [releases] page.
- All [releases] are signed with
  `EBE5 EBC0 F49E 38A6 9FC7 EA26 7366 AE4A 6774 8172`
  ([keybase.io/fleshgrinder]).

<!-- @formatter:off -->
[CodeCov]: https://codecov.io/gh/Fleshgrinder/jvm-platform
[CONTRIBUTING.md]: https://github.com/Fleshgrinder/.github/blob/main/CONTRIBUTING.md
[gradle-exe-plugin]: https://github.com/Fleshgrinder/gradle-exe-plugin
[keep a changelog]: https://keepachangelog.com/
[keybase.io/fleshgrinder]: https://keybase.io/fleshgrinder
[Maven Central]: https://search.maven.org/artifact/com.fleshgrinder/jvm-platform
[musl]: https://musl.libc.org/
[releases]: https://github.com/Fleshgrinder/jvm-platform/releases
[semantic versioning]: http://semver.org/
[UNLICENSE.md]: UNLICENSE.md
[Unlicense]: https://unlicense.org/
[Usage.java]: src/test/java/com/fleshgrinder/platform/Usage.java


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/), and this
project adheres to [Semantic Versioning](https://semver.org/).

## [Unreleased]
### Added
- initial release

[Unreleased]: https://github.com/Fleshgrinder/jvm-platform


src/test/java/com/fleshgrinder/platform/Usage.java
package com.fleshgrinder.platform;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Usage Examples
 *
 * @see #platform()
 * @see #os()
 * @see #arch()
 * @see #env()
 */
final class Usage {
    /**
     * The {@link Platform} is the main entry point and combined {@link Os} and
     * {@link Arch}. It does not contain the {@link Env} because it is required
     * only is special situations and expensive to determine compared to the
     * others.
     */
    @Test void platform() {
        // current() gets the Os and Arch of this JVM
        final Platform current = Platform.current();
        System.out.println(current.os);
        System.out.println(current.arch);
        System.out.println(current.id);

        // parse() is able to extract platform information
        final Platform custom = Platform.parse("Command-Linux-Aarch64_BE.exe");
        assertEquals(Os.LINUX, custom.os);
        assertEquals(Arch.ARM_64_BE, custom.arch);
        // getId() is strictly defined and parser friendly, refer to the Javadoc
        // of it for detailed information
        assertEquals("linux-arm-64-be", custom.id);

        // it is also possible to construct any kind of platform we want
        System.out.println(new Platform(Os.LINUX, Arch.ALPHA_64));
        System.out.println(new Platform(Os.WINDOWS));
        System.out.println(new Platform(Arch.SPARC_64));
    }

    @Test void os() {
        System.out.println(Os.current());

        assertEquals(Os.UNKNOWN, Os.parse("PSP"));
        assertEquals(Os.DARWIN, Os.parse("Mac OS X"));

        assertEquals("", Os.LINUX.getExecutableExtension());
        assertEquals(".exe", Os.WINDOWS.getExecutableExtension());

        assertEquals("cmd", Os.LINUX.withExecutableExtension("cmd"));
        assertEquals("cmd.exe", Os.WINDOWS.withExecutableExtension("cmd"));

        assertEquals(".so", Os.LINUX.getLinkLibraryExtension());
        assertEquals(".lib", Os.WINDOWS.getLinkLibraryExtension());

        assertEquals("some.so", Os.LINUX.withLinkLibraryExtension("some"));
        assertEquals("some.lib", Os.WINDOWS.withLinkLibraryExtension("some"));

        assertEquals(".dylib", Os.DARWIN.getSharedLibraryExtension());
        assertEquals(".so", Os.LINUX.getSharedLibraryExtension());
        assertEquals(".dll", Os.WINDOWS.getSharedLibraryExtension());

        assertEquals("some.dylib", Os.DARWIN.withSharedLibraryExtension("some"));
        assertEquals("some.so", Os.LINUX.withSharedLibraryExtension("some"));
        assertEquals("some.dll", Os.WINDOWS.withSharedLibraryExtension("some"));

        assertEquals(".a", Os.LINUX.getStaticLibraryExtension());
        assertEquals(".lib", Os.WINDOWS.getStaticLibraryExtension());

        assertEquals("some.a", Os.LINUX.withStaticLibraryExtension("some"));
        assertEquals("some.lib", Os.WINDOWS.withStaticLibraryExtension("some"));
    }

    @Test void arch() {
        System.out.println(Arch.current());

        assertEquals(Arch.UNKNOWN_UNKNOWN, Arch.parse("test"));
        assertEquals(Arch.UNKNOWN_32, Arch.parse("test32"));
        assertEquals(Arch.UNKNOWN_64, Arch.parse("test64"));
        assertEquals(Arch.ARM_64, Arch.parse("aarch64"));

        assertEquals(32, Arch.X86_32.getBitness());
        assertEquals(64, Arch.X86_64.getBitness());
        assertTrue(Arch.ARM_32.is32bit());
        assertTrue(Arch.ARM_64.is64bit());
        assertTrue(Arch.ARM_64_BE.isArm());
        assertFalse(Arch.ARM_64_BE.isArmLe());

        assertTrue(Arch.ARM_32.isArm());
        assertTrue(Arch.ARM_32.isArmLe());
        assertFalse(Arch.ARM_32.isArmBe());
    }

    @Test void env() {
        System.out.println(Env.current());

        assertEquals(Env.UNKNOWN, Env.parse("elf"));
        assertEquals(Env.GLIBC, Env.parse("gcc"));
        assertEquals(Env.MUSL, Env.parse("musl"));
        assertEquals(Env.MSVC, Env.parse("win"));
    }
}


src/test/java/com/fleshgrinder/platform/OsTest.java
package com.fleshgrinder.platform;

import java.io.File;
import java.util.function.Function;
import java.util.function.Supplier;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.EnumSource;
import org.junitpioneer.jupiter.ClearSystemProperty;
import org.junitpioneer.jupiter.SetSystemProperty;

import static com.fleshgrinder.platform.SystemProperties.FS_SEP;
import static com.fleshgrinder.platform.SystemProperties.OS_NAME;
import static com.fleshgrinder.platform.SystemProperties.VM_NAME;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.params.provider.EnumSource.Mode.EXCLUDE;

final class OsTest {
    @ParameterizedTest
    @ClearSystemProperty(key = OS_NAME)
    @ClearSystemProperty(key = FS_SEP)
    @ClearSystemProperty(key = VM_NAME)
    @CsvSource({
        "UNKNOWN, ''",
        "UNKNOWN, '    '",
        "UNKNOWN, '\t\n\t\n'",
        "UNKNOWN, rv32imac",
        "UNKNOWN, rv64imac",
        "UNKNOWN, os4000",
        "AIX, AIX",
        "ANDROID, Android",
        "DARWIN, Darwin",
        "DARWIN, Mac OS X",
        "DARWIN, Mac OS",
        "DRAGONFLYBSD, dragonflybsd",
        "FREEBSD, freebsd",
        "FUCHSIA, fuchsia",
        "HAIKU, haiku",
        "HPUX, hpux",
        "IBMI, ibmi",
        "IBMI, os400",
        "ILLUMOS, illumos",
        "LINUX, linux",
        "NETBSD, netbsd",
        "OPENBSD, openbsd",
        "PLAN9, plan9",
        "QNX, procnto",
        "QNX, qnx",
        "REDOX, redox",
        "SOLARIS, solaris",
        "SOLARIS, sunos",
        "VXWORKS, vxworks",
        "ZOS, zos",

        "AIX, AIX",
        "AIX, garbage aix garbage",
        "ANDROID, ANDROID",
        "ANDROID, garbage android garbage",
        "ANDROID, linux android linux",
        "DARWIN, DARWIN",
        "DARWIN, garbage apple garbage",
        "DARWIN, garbage darwin garbage",
        "DARWIN, garbage mac garbage",
        "DARWIN, garbage mac os garbage",
        "DARWIN, garbage mac os x garbage",
        "DARWIN, garbage mac osx garbage",
        "DARWIN, garbage macos garbage",
        "DARWIN, garbage macos x garbage",
        "DARWIN, garbage macosx garbage",
        "DRAGONFLYBSD, DRAGONFLYBSD",
        "DRAGONFLYBSD, garbage dragonfly garbage",
        "FREEBSD, FREEBSD",
        "FREEBSD, garbage freebsd garbage",
        "FUCHSIA, FUCHSIA",
        "FUCHSIA, garbage fuchsia garbage",
        "HAIKU, garbage haiku garbage",
        "HAIKU, HAIKU",
        "HPUX, garbage hp-ux garbage",
        "HPUX, garbage hpux garbage",
        "HPUX, HPUX",
        "IBMI, garbage ibm i garbage",
        "IBMI, garbage ibmi garbage",
        "IBMI, garbage os400 garbage",
        "IBMI, IBMI",
        "ILLUMOS, garbage illum os garbage",
        "ILLUMOS, garbage illumos garbage",
        "ILLUMOS, ILLUMOS",
        "LINUX, garbage nix garbage",
        "LINUX, garbage nux garbage",
        "LINUX, LINUX",
        "NETBSD, garbage netbsd garbage",
        "NETBSD, NETBSD",
        "OPENBSD, garbage openbsd garbage",
        "OPENBSD, OPENBSD",
        "PLAN9, garbage plan9 garbage",
        "PLAN9, PLAN9",
        "QNX, garbage procnto garbage",
        "QNX, garbage qnx garbage",
        "QNX, QNX",
        "REDOX, garbage redox garbage",
        "REDOX, REDOX",
        "SOLARIS, garbage solaris garbage",
        "SOLARIS, garbage sun os garbage",
        "SOLARIS, garbage sunos garbage",
        "SOLARIS, SOLARIS",
        "VXWORKS, garbage VxWorks garbage",
        "VXWORKS, VXWORKS",
        "WINDOWS, garbage w32 garbage",
        "WINDOWS, garbage w64 garbage",
        "WINDOWS, garbage win32 garbage",
        "WINDOWS, garbage windows garbage",
        "WINDOWS, Windows 10",
        "WINDOWS, Windows 7",
        "WINDOWS, Windows XP",
        "WINDOWS, WINDOWS",
        "ZOS, garbage z/OS garbage",
        "ZOS, garbage zos garbage",
        "ZOS, z/OS",
        "ZOS, ZOS",
    })
    void current(final @NotNull Os expected, final @NotNull String osName) {
        SystemProperties.set(OS_NAME, osName);
        assertEquals(expected, Os.current());
    }

    @SetSystemProperty(key = OS_NAME, value = "linux")
    @SetSystemProperty(key = VM_NAME, value = "Dalvik")
    @ClearSystemProperty(key = FS_SEP)
    @Test void currentAndroid() {
        assertEquals(Os.ANDROID, Os.current());
    }

    @SetSystemProperty(key = FS_SEP, value = "\\")
    @ClearSystemProperty(key = VM_NAME)
    @ClearSystemProperty(key = OS_NAME)
    @Test void currentWindows() {
        assertEquals(Os.WINDOWS, Os.current());
    }

    private static void assertExtension(
        final @NotNull String ext,
        final @NotNull Supplier<String> getExt,
        final @NotNull Function<String, String> withExtString,
        final @NotNull Function<File, File> withExtFile
    ) {
        assertAll(
            () -> assertEquals(ext, getExt.get()),
            () -> assertEquals("some/path" + ext, withExtString.apply("some/path")),
            () -> assertEquals(new File("some/path" + ext), withExtFile.apply(new File("some/path")))
        );
    }

    @Test void executableExtensionWindows() {
        assertExtension(
            ".exe",
            Os.WINDOWS::getExecutableExtension,
            Os.WINDOWS::withExecutableExtension,
            Os.WINDOWS::withExecutableExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"WINDOWS"}, mode = EXCLUDE)
    void executableExtension(final @NotNull Os os) {
        assertExtension(
            "",
            os::getExecutableExtension,
            os::withExecutableExtension,
            os::withExecutableExtension
        );
    }

    @Test void linkLibraryExtensionWindows() {
        assertExtension(
            ".lib",
            Os.WINDOWS::getLinkLibraryExtension,
            Os.WINDOWS::withLinkLibraryExtension,
            Os.WINDOWS::withLinkLibraryExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"WINDOWS"}, mode = EXCLUDE)
    void linkLibraryExtension(final @NotNull Os os) {
        assertExtension(
            ".so",
            os::getLinkLibraryExtension,
            os::withLinkLibraryExtension,
            os::withLinkLibraryExtension
        );
    }

    @Test void sharedLibraryExtensionWindows() {
        assertExtension(
            ".dll",
            Os.WINDOWS::getSharedLibraryExtension,
            Os.WINDOWS::withSharedLibraryExtension,
            Os.WINDOWS::withSharedLibraryExtension
        );
    }

    @Test void sharedLibraryExtensionDarwin() {
        assertExtension(
            ".dylib",
            Os.DARWIN::getSharedLibraryExtension,
            Os.DARWIN::withSharedLibraryExtension,
            Os.DARWIN::withSharedLibraryExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"DARWIN", "WINDOWS"}, mode = EXCLUDE)
    void sharedLibraryExtension(final @NotNull Os os) {
        assertExtension(
            ".so",
            os::getSharedLibraryExtension,
            os::withSharedLibraryExtension,
            os::withSharedLibraryExtension
        );
    }

    @Test void staticLibraryExtensionWindows() {
        assertExtension(
            ".lib",
            Os.WINDOWS::getStaticLibraryExtension,
            Os.WINDOWS::withStaticLibraryExtension,
            Os.WINDOWS::withStaticLibraryExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"WINDOWS"}, mode = EXCLUDE)
    void staticLibraryExtension(final @NotNull Os os) {
        assertExtension(
            ".a",
            os::getStaticLibraryExtension,
            os::withStaticLibraryExtension,
            os::withStaticLibraryExtension
        );
    }
}


src/main/java/com/fleshgrinder/platform/Env.java
package com.fleshgrinder.platform;

import java.io.InputStream;
import java.util.concurrent.TimeUnit;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.VisibleForTesting;

import static com.fleshgrinder.platform.Utils.id;
import static com.fleshgrinder.platform.Utils.normalize;

/**
 * Environment
 *
 * <p>A platform's environment defines the ABI/libc requirements of shared
 * libraries or dynamic executables.
 *
 * @see <a href="https://en.wikipedia.org/wiki/C_standard_library">Wikipedia</a>
 */
@SuppressWarnings("SpellCheckingInspection")
public enum Env {
    UNKNOWN,

    /**
     * Bionic
     *
     * @see <a href="https://en.wikipedia.org/wiki/Bionic_(software)">Wikipedia</a>
     */
    BIONIC,

    /**
     * BSD libc
     *
     * @see <a href="https://en.wikipedia.org/wiki/BSD_libc">Wikipedia</a>
     */
    BSDLIBC,

    /**
     * dietlibc
     *
     * @see <a href="https://en.wikipedia.org/wiki/Dietlibc">Wikipedia</a>
     */
    DIETLIBC,

    /**
     * GNU C Library
     *
     * @see <a href="https://en.wikipedia.org/wiki/GNU_C_Library">Wikipedia</a>
     */
    GLIBC,

    /**
     * klibc
     *
     * @see <a href="https://en.wikipedia.org/wiki/Klibc">Wikipedia</a>
     */
    KLIBC,

    /**
     * Microsoft Visual C++
     *
     * @see <a href="https://en.wikipedia.org/wiki/Microsoft_Visual_C++">Wikipedia</a>
     */
    MSVC,

    /**
     * musl
     *
     * @see <a href="https://en.wikipedia.org/wiki/Musl">Wikipedia</a>
     */
    MUSL,

    /**
     * Newlib
     *
     * @see <a href="https://en.wikipedia.org/wiki/Newlib">Wikipedia</a>
     */
    NEWLIB,

    /**
     * µClibc
     *
     * @see <a href="https://en.wikipedia.org/wiki/UClibc">Wikipedia</a>
     * @see <a href="https://uclibc-ng.org/">uClibc-ng</a>
     */
    UCLIBC;

    /**
     * {@link #name()} in {@code lower-dash-case}
     */
    public final @NotNull String id = id(name());

    /**
     * Gets the platform's env.
     *
     * <p>Determining the env is expensive compared to {@link Arch#current()} or
     * {@link Os#current()} because a process must be spawned to retrieve
     * information regarding this platform if it is not {@link Os#WINDOWS}. On
     * {@link Os#WINDOWS} the result always is {@link #MSVC}.
     *
     * @return current env or {@link #UNKNOWN}
     */
    @Contract(pure = true)
    public static @NotNull Env current() {
        return current(Os.current(), null);
    }

    /**
     * Parses the given value and tries to match it to one of the known envs.
     *
     * @param value to parse
     * @return matching env or {@link #UNKNOWN}
     * @throws NullPointerException if value is {@code null}
     */
    @Contract(pure = true)
    public static @NotNull Env parse(final @NotNull CharSequence value) {
        if (value.length() > 0) {
            final String it = normalize(value);
            for (final Env env : Env.values()) if (env != Env.UNKNOWN && it.contains(env.id)) return env;
            // Matching `gnu` is dangerous because of the GNU operating systems
            // that sometimes identify themselves only with GNU as well, but we
            // need `gnu` here because of gcc. Future has to show if this is a
            // real world problem, or not.
            //
            // https://en.wikipedia.org/wiki/GNU_variants
            if (it.matches(".*g(cc|nu).*")) return GLIBC;
            if (it.matches(".*(apple|bsd|darwin|mac|osx|ios|dragonfly).*")) return BSDLIBC;
            if (it.matches(".*(crtdll|ucrt|vcruntime|vs|win).*")) return MSVC;
            if (it.contains("android")) return BIONIC;
        }
        return UNKNOWN;
    }

    @Contract(pure = true)
    @VisibleForTesting
    static @NotNull Env current(final @NotNull Os os, final @Nullable String lddPath) {
        switch (os) {
            case ANDROID:
                return BIONIC;
            case DARWIN:
            case DRAGONFLYBSD:
            case FREEBSD:
            case NETBSD:
            case OPENBSD:
                return BSDLIBC;
            case WINDOWS:
                return MSVC;
        }
        return parse(ldd(lddPath));
    }

    /**
     * @return first lowercased line from {@code ldd --version}, or an empty
     *     string if execution fails (e.g., {@code ldd} is not in {@code PATH})
     */
    @Contract(pure = true)
    @VisibleForTesting
    static @NotNull String ldd(final @Nullable String lddPath) {
        final @NotNull Process p;
        try {
            p = new ProcessBuilder(lddPath == null ? "ldd" : lddPath, "--version").start();
        } catch (final Throwable ignored) {
            return "";
        }

        final @NotNull StringBuilder sb = new StringBuilder(40);
        try {
            if (p.waitFor(1, TimeUnit.SECONDS)) {
                try (final @NotNull InputStream s = p.getInputStream()) {
                    int c;
                    while ((c = s.read()) != -1 && c != '\n') {
                        if ('A' <= c && c <= 'Z') sb.append((char) (c + 32));
                        else sb.append((char) c);
                    }
                }
            }
        } catch (final Throwable ignored) {
            // fall through
        }

        p.destroyForcibly(); // closes all streams
        return sb.toString();
    }

    @Override public String toString() {
        return id;
    }
}


src/main/java/com/fleshgrinder/platform/Utils.java
package com.fleshgrinder.platform;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

final class Utils {
    private Utils() {}

    /**
     * Transforms {@code UPPER_SNAKE_CASE} to {@code lower-dash-case}.
     */
    @Contract(pure = true)
    static @NotNull String id(final @NotNull String chars) {
        final int l = chars.length();
        final StringBuilder sb = new StringBuilder(l);
        int i = 0;
        do {
            final char c = chars.charAt(i);
            if (c == '_') sb.append('-');
            else if ('A' <= c && c <= 'Z') sb.append((char) (c + 32));
            else sb.append(c);
        } while (++i < l);
        return sb.toString();
    }

    /**
     * Normalizes the given chars by lowering all upper chars and replacing all
     * non-alphanumeric chars with a dash ({@code -}).
     *
     * @param chars to normalize
     * @return normalized string
     * @throws NullPointerException if {@code chars} is {@code null}
     * @see #normalize(CharSequence, boolean)
     */
    @Contract(pure = true)
    static @NotNull String normalize(final @NotNull CharSequence chars) {
        return normalize(chars, false);
    }

    /**
     * Normalizes the given chars by lowering all upper chars and either
     * stripping or replacing all non-alphanumeric chars.
     *
     * @param chars to normalize
     * @param strip non-alphanumeric chars ({@code true}) or replace them with a
     *     dash ({@code -}, {@code false})
     * @return normalized string
     * @throws NullPointerException if {@code chars} is {@code null}
     */
    @Contract(pure = true)
    static @NotNull String normalize(final @NotNull CharSequence chars, final boolean strip) {
        final int l = chars.length();
        if (l == 0) return "";
        final StringBuilder sb = new StringBuilder(l);
        int i = 0;
        do {
            final char c = chars.charAt(i);
            if (('0' <= c && c <= '9') || ('a' <= c && c <= 'z')) sb.append(c);
            else if ('A' <= c && c <= 'Z') sb.append((char) (c + 32));
            else if (!strip) sb.append('-');
        } while (++i < l);
        return sb.toString();
    }
}


src/main/java/com/fleshgrinder/platform/Os.java
package com.fleshgrinder.platform;

import java.io.File;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import static com.fleshgrinder.platform.Utils.id;
import static com.fleshgrinder.platform.Utils.normalize;

/**
 * Operating System
 *
 * @see <a href="https://en.wikipedia.org/wiki/Operating_system">Wikipedia</a>
 */
@SuppressWarnings("SpellCheckingInspection")
public enum Os {
    UNKNOWN,

    /**
     * IBM AIX
     *
     * @see <a href="https://en.wikipedia.org/wiki/IBM_AIX">Wikipedia</a>
     */
    AIX,

    /**
     * Google Android
     *
     * @see <a href="https://en.wikipedia.org/wiki/Android_(operating_system)">Wikipedia</a>
     */
    ANDROID,

    /**
     * Apple Darwin (Mac OS X)
     *
     * @see <a href="https://en.wikipedia.org/wiki/Darwin">Wikipedia</a>
     */
    DARWIN,

    /**
     * DragonFly BSD
     *
     * @see <a href="https://en.wikipedia.org/wiki/DragonFly_BSD">Wikipedia</a>
     */
    DRAGONFLYBSD,

    /**
     * FreeBSD
     *
     * @see <a href="https://en.wikipedia.org/wiki/FreeBSD">Wikipedia</a>
     */
    FREEBSD,

    /**
     * Google Fuchsia
     *
     * @see <a href="https://en.wikipedia.org/wiki/Google_Fuchsia">Wikipedia</a>
     */
    FUCHSIA,

    /**
     * Haiku
     *
     * @see <a href="https://en.wikipedia.org/wiki/Haiku_(operating_system)">Wikipedia</a>
     */
    HAIKU,

    /**
     * HP-UX
     *
     * @see <a href="https://en.wikipedia.org/wiki/HP-UX">Wikipedia</a>
     */
    HPUX,

    /**
     * Linux
     *
     * @see <a href="https://en.wikipedia.org/wiki/Linux">Wikipedia</a>
     */
    LINUX,

    /**
     * IBM i
     *
     * @see <a href="https://en.wikipedia.org/wiki/IBM_i">Wikipedia</a>
     */
    IBMI,

    /**
     * illumos
     *
     * @see <a href="https://en.wikipedia.org/wiki/Illumos">Wikipedia</a>
     */
    ILLUMOS,

    /**
     * NetBSD
     *
     * @see <a href="https://en.wikipedia.org/wiki/NetBSD">Wikipedia</a>
     */
    NETBSD,

    /**
     * OpenBSD
     *
     * @see <a href="https://en.wikipedia.org/wiki/OpenBSD">Wikipedia</a>
     */
    OPENBSD,

    /**
     * Plan9 from Bell Labs
     *
     * @see <a href="https://en.wikipedia.org/wiki/Plan_9_from_Bell_Labs">Wikipedia</a>
     * @see <a href="https://p9f.org/">p9f.org</a>
     */
    PLAN9,

    /**
     * QNX
     *
     * @see <a href="https://en.wikipedia.org/wiki/QNX">Wikipedia</a>
     */
    QNX,

    /**
     * Redox
     *
     * @see <a href="https://en.wikipedia.org/wiki/Redox_(operating_system)">Wikipedia</a>
     */
    REDOX,

    /**
     * Oracle Solaris (SunOS)
     *
     * @see <a href="https://en.wikipedia.org/wiki/Oracle_Solaris">Wikipedia</a>
     */
    SOLARIS,

    /**
     * VxWorks
     *
     * @see <a href="https://en.wikipedia.org/wiki/VxWorks">Wikipedia</a>
     */
    VXWORKS,

    /**
     * Microsoft Windows
     *
     * @see <a href="https://en.wikipedia.org/wiki/Microsoft_Windows">Wikipedia</a>
     */
    WINDOWS,

    /**
     * z/OS
     *
     * @see <a href="https://en.wikipedia.org/wiki/Z/OS">Wikipedia</a>
     */
    ZOS;

    /**
     * {@link #name()} in {@code lower-dash-case}
     */
    public final @NotNull String id = id(name());

    /**
     * Determines if the current OS is Windows.
     *
     * @return {@code true} if the current OS is Windows.
     */
    @Contract(pure = true)
    static boolean isWindows() {
        return "\\".equals(System.getProperty("file.separator"));
    }

    /**
     * Gets this platform's operating system.
     *
     * @return current OS or {@link #UNKNOWN}
     */
    @Contract(pure = true)
    public static @NotNull Os current() {
        if (isWindows()) return WINDOWS;
        final @NotNull String it = System.getProperty("os.name", "");
        switch (normalize(it, true)) {
            case "aix":
                return AIX;
            case "android":
                return ANDROID;
            case "linux":
                return "dalvik".equalsIgnoreCase(System.getProperty("java.vm.name", "")) ? ANDROID : LINUX;
            case "darwin":
            case "macosx":
            case "macos":
                return DARWIN;
            case "dragonflybsd":
                return DRAGONFLYBSD;
            case "freebsd":
                return FREEBSD;
            case "fuchsia":
                return FUCHSIA;
            case "haiku":
                return HAIKU;
            case "hpux":
                return HPUX;
            case "os400":
                return IBMI;
            case "illumos":
                return ILLUMOS;
            case "netbsd":
                return NETBSD;
            case "openbsd":
                return OPENBSD;
            case "plan9":
                return PLAN9;
            case "qnx":
            case "procnto":
                return QNX;
            case "redox":
                return REDOX;
            case "solaris":
            case "sunos":
                return SOLARIS;
            case "vxworks":
                return VXWORKS;
            case "zos":
                return ZOS;
        }
        return parse(it);
    }

    /**
     * Parses the given value and tries to extract
     *
     * @param value to parse
     * @return matching OS or {@link #UNKNOWN}
     * @throws NullPointerException if value is {@code null}
     */
    @Contract(pure = true)
    public static @NotNull Os parse(final @NotNull CharSequence value) {
        if (value.length() > 0) {
            final @NotNull String it = normalize(value);
            // Android MUST come before Linux because they often come together
            if (it.contains("android")) return ANDROID;
            if (it.matches(".*n[iu]x.*")) return LINUX;
            // (32|64)imac as in RV(32|64)IMAC ISA may be embedded in a platform
            // identifier but is not necessarily targeting Darwin.
            // `osx` is considered Darwin even standalone but `os-x` is not,
            // because it would also match `os-x86` which leads to too many
            // false positives.
            if (it.matches(".*(apple|darwin|(?<!(32|64)i)mac|osx|ios).*")) return DARWIN;
            // MUST come after darWIN!
            // w32 and w64 are from gcc (actually mingw-w64)
            if (it.matches(".*w(in|32|64).*")) return WINDOWS;
            if (it.contains("aix")) return AIX;
            if (it.contains("dragonfly")) return DRAGONFLYBSD;
            if (it.contains("freebsd")) return FREEBSD;
            if (it.contains("fuchsia")) return FUCHSIA;
            if (it.contains("haiku")) return HAIKU;
            if (it.contains("netbsd")) return NETBSD;
            if (it.contains("openbsd")) return OPENBSD;
            if (it.contains("plan9")) return PLAN9;
            if (it.contains("redox")) return REDOX;
            if (it.contains("vxworks")) return VXWORKS;
            if (it.matches(".*hp-?ux.*")) return HPUX;
            if (it.matches(".*(ibm-?i|os400[^0-9]).*")) return IBMI;
            if (it.matches(".*illum-?os.*")) return ILLUMOS;
            if (it.matches(".*(qnx|procnto).*")) return QNX;
            if (it.matches(".*(s(olaris|un-?os)).*")) return SOLARIS;
            if (it.matches(".*z-?os.*")) return ZOS;
        }
        return UNKNOWN;
    }

    /**
     * @return extension for executables with leading period or an empty string
     *     if not applicable on this platform
     */
    @Contract(pure = true)
    public @NotNull String getExecutableExtension() {
        if (this == WINDOWS) return ".exe";
        return "";
    }

    /**
     * @param path to append the extension to
     * @return path with {@link #getExecutableExtension}
     */
    @Contract(pure = true)
    public @NotNull String withExecutableExtension(final @NotNull String path) {
        return path + getExecutableExtension();
    }

    /**
     * @param file to append the extension to
     * @return file with {@link #getExecutableExtension}
     */
    @Contract(pure = true)
    public @NotNull File withExecutableExtension(final @NotNull File file) {
        return new File(file.getParentFile(), withExecutableExtension(file.getName()));
    }

    /**
     * @return extension for link libraries with leading period on this platform
     */
    @Contract(pure = true)
    public @NotNull String getLinkLibraryExtension() {
        if (this == WINDOWS) return ".lib";
        return ".so";
    }

    /**
     * @param path to append the extension to
     * @return path with {@link #getLinkLibraryExtension}
     */
    @Contract(pure = true)
    public @NotNull String withLinkLibraryExtension(final @NotNull String path) {
        return path + getLinkLibraryExtension();
    }

    /**
     * @param file to append the extension to
     * @return file with {@link #getLinkLibraryExtension}
     */
    @Contract(pure = true)
    public @NotNull File withLinkLibraryExtension(final @NotNull File file) {
        return new File(file.getParentFile(), withLinkLibraryExtension(file.getName()));
    }

    /**
     * @return extension for shared libraries with leading period on this
     *     platform
     */
    @Contract(pure = true)
    public @NotNull String getSharedLibraryExtension() {
        switch (this) {
            case WINDOWS:
                return ".dll";
            case DARWIN:
                return ".dylib";
        }
        return ".so";
    }

    /**
     * @param path to append the extension to
     * @return path with {@link #getSharedLibraryExtension}
     */
    @Contract(pure = true)
    public @NotNull String withSharedLibraryExtension(final @NotNull String path) {
        return path + getSharedLibraryExtension();
    }

    /**
     * @param file to append the extension to
     * @return file with {@link #getSharedLibraryExtension}
     */
    @Contract(pure = true)
    public @NotNull File withSharedLibraryExtension(final @NotNull File file) {
        return new File(file.getParentFile(), withSharedLibraryExtension(file.getName()));
    }

    /**
     * @return extension for static libraries with leading period on this
     *     platform
     */
    @Contract(pure = true)
    public @NotNull String getStaticLibraryExtension() {
        if (this == WINDOWS) return ".lib";
        return ".a";
    }

    /**
     * @param path to append the extension to
     * @return path with {@link #getStaticLibraryExtension}
     */
    @Contract(pure = true)
    public @NotNull String withStaticLibraryExtension(final @NotNull String path) {
        return path + getStaticLibraryExtension();
    }

    /**
     * @param file to append the extension to
     * @return file with {@link #getStaticLibraryExtension}
     */
    @Contract(pure = true)
    public @NotNull File withStaticLibraryExtension(final @NotNull File file) {
        return new File(file.getParentFile(), withStaticLibraryExtension(file.getName()));
    }

    @Override public String toString() {
        return id;
    }
}
